### DISCUSSION TITLE
Prisam transactions, how to implement independent writes with Optimistic Currency Control

### DISCUSSION THREAD

**Discussion by doddi321** [prisma/prisma#7257] [Q&A] [answered]
I have been playing around with Prisma for a bit and got to the point where I needed to do a transaction where the model/table being updated relies on data that was read previously in the application from another model/table and the update should only happen if that read data has not changed. 

So my question is how do I implement this with Prisma?

To be more specific I have the two following models:
```prisma
model Invitation {
  id        Int      @id @default(autoincrement())
  code      String   @unique @db.VarChar
  maxPeople Int      @map("max_people")
  createdAt DateTime @default(now()) @map("created_at") @db.Timestamp(6)
  version   Int      @default(0)
  users     User[]

  @@map("invitations")
}

model User {
  id             String           @id @default(dbgenerated("uuid_generate_v4()")) @db.Uuid
  count_         Int              @default(autoincrement()) @map("count")
  createdAt      DateTime         @default(now()) @map("created_at") @db.Timestamp(6)
  name           String           @db.VarChar
  password       String           @db.VarChar
  email          String           @db.VarChar
  googleId       String?          @map("google_id") @db.VarChar
  imageUrl       String?          @map("image_url") @db.VarChar
  points         Int              @default(0)
  invitationCode String?          @map("invitation_code") @db.VarChar
  invitation     Invitation?  @relation(fields: [invitationCode], references: [code])
  productHistory ProductHistory[]

  @@unique([email])
  @@map("users")
}
```

and what I want to do is create another user with a certain invitation code if and only if the _maxPeople_ field in the **Invitation** model is not equal or greaten then the amount of already connected _users_

So the general flow of the application is: 

1. Retrieve the **Invitation** object using a `code` 
2. check if the `maxPeople` field has been exceeded by the connected _users_, if so then throw an exception else continue
3. proceed to create the **User** 

However this is execution plan is somewhat flawed since someone might have connected a **User** to the **Invitation** object after the object was fetched from the database and thus my application will determine that the  `maxPeople`  field has not been exceeded event though it indeed has. 

My initial plan was to use Optimistic Concurrency Control as specified here: (https://www.prisma.io/docs/guides/performance-and-optimization/prisma-client-transactions-guide#optimistic-concurrency-control)  and add a version field to the **Invitation** object to determine if a user has been added after the fetch but I then need to create the user so I would somehow need to both update the version and create the user in a transaction like so: 

```ts
/* user.repository.ts */
 async createUser(user: User): Promise<string> {
    const [createdUser] = await this.prisma.$transaction([
      this.createUserOperation(user),
      this.invitationRepository.incrementInvitation_OCC_Version(
        user.invitation,
      ),
    ]);
    return createdUser.id;
  }

 private createUserOperation(user: User) {
    return this.prisma.user.create({
      data: {
        email: user.email,
        name: user.name,
        password: user.hashedPassword,
        invitationCode: user.invitation.code,
      },
    });
  }
```
```ts
 /*  invitation.repository.ts */
 incrementInvitation_OCC_Version(invitation: Invitation) {
    return this.prisma.invitationCode.updateMany({
      data: {
        version: {
          increment: 1,
        },
      },
      where: {
        code: invitation.code,
        version: invitation.version,
      },
    });
```

but the problem with this is that _updateMany_ doesn't throw an exception if there were no rows updated so if the version had changed then the user would still be created. 

Does anyone have a solution to my problem? Any help would be greatly appreaciated. 

Here: https://www.prisma.io/docs/guides/performance-and-optimization/prisma-client-transactions-guide/#conclusion it says that "If you feel like you have use cases in your application that are not covered by any of the suggested options, please open a GitHub issue" but this felt more like a question then an issue, if you prefer me to open this up as an issue then I would be happy to.

**Comment by ryands17** [comment, marked as answer]
@doddi321 👋 

Updating both in the transaction is not necessary. As shown in the docs, booking a seat with versioning is done without the use of transactions. 

In the above case, you can just use `updateMany` without wrapping it in a transaction and check if the `count` returned is **0**, it means that there was a version mismatch and you can throw an error as mentioned in the example:

![image](https://user-images.githubusercontent.com/19697099/119454529-3fca6d00-bd56-11eb-838a-db2072506194.png)

Let me know if this helps 👍

**Reply by doddi321** [reply]
yeah but then I need to create a user, so what about the followin scenario:
1. check if the versioning is the same
2. then somebody is able to sneek in and update the version
3. I create the user

so my problem is that I need to check/update the versioning and create the user together, that is that the two operations fail or succeed together.

**Reply by ryands17** [reply]
In this case, I don't see any other way apart from a [raw query](https://www.prisma.io/docs/concepts/components/prisma-client/raw-database-access). So you would need to write the query with `begin` and `commit` directly.

As for the issue, we will be introducing long running transactions as mentioned here: https://github.com/prisma/prisma/issues/1844#issuecomment-846477520 so your use case will be covered by this in the near future :)

**Reply by doddi321** [reply]
ah I see, thanks for the quick replies, greatly appreciated.
